### ISSUE TITLE
Value is not being deeply decoded before being passed to `TUnion` decoder

### ISSUE THREAD

**Issue by EmilEriksen** [sinclairzx81/typebox#676]
I apologise if I'm butchering the terminology. I'm trying to define a transform on a union schema. This union schema has a property with a date schema that also defines a transform i.e.
```typescript
const dateSchema = Type.Transform(Type.String())
  .Decode((ds: string): Date => new Date(ds))
  .Encode((d: Date): string => d.toISOString());

const complexObjectSchema = Type.Union([
  Type.Object({
    date: dateSchema,
  }),
  Type.Number(),
]);

const transformComplexSchema = Type.Transform(complexObjectSchema)
  .Decode((o) => {
    if (typeof o === "object") {
      // date is supposed to be a Date according to type inference, but it's a string when it's logged.
      console.log("union", typeof o.date); // string
    }

    return o;
  })
  .Encode((o) => o);
```

If I define a transform like this
```typescript
Type.Transform(Type.Object({ date: dateSchema }))
  .Decode((o) => {
    console.log("simple", typeof o.date); // object

    return o;
  })
  .Encode((o) => o);
```
`date` is a `Date`.

I've created a [CodeSandbox](https://codesandbox.io/p/sandbox/modest-sunset-pkcsr5) to illustrate the issue

**Comment by sinclairzx81** [comment]
@EmilEriksen Hi, thanks for reporting. 

This looks like a bug (although it may be to do with TypeBox establishing a correct transform order when transforming unions (where interior types should transform before outer types if possible)). Both union and intersect types have been difficult to get right as limitations and ambiguities can pop-up depending on the composed type, it's been challenging.

Ill take a look at this as soon as I can, unfortunately, I'm currently investigating some compiler issues that have popped up the TS 5.4-dev nightly builds which is taking a bit of effort to resolve (and compiler related issues take priority over some of the runtime functionality). Depending on how things go, will try and resolve this by the weekend. Will flag as a bug and pick this up as soon as I can.

As a temporary workaround, if you can avoid applying transforms directly to Union and Intersect types, you may have better luck (as this issue applies specifically to correctly applying the transform to these types, but non-composite types should be ok)

Cheers
S

**Comment by EmilEriksen** [comment]
Hi @sinclairzx81 

Thanks for getting back to me so quickly. I can imagine that it's pretty difficult to get right. It's a very useful feature though. Being able to define how to encode and decode a type along with the definition is very powerful for ensuring it's done consistently throughout your codebase.

No rush from me personally - i already have a workaround :)

Have a great day!

**Comment by sinclairzx81** [comment]
@EmilEriksen Hi, Apologies for the delay on this. 

Just to keep you updated, I'm going to be fixing this issue on the next minor revision of TypeBox (0.32.0), along with bringing several new enhancements to the library. I've setup a PR https://github.com/sinclairzx81/typebox/pull/689 which includes the fix. I'm aiming to get this published within the next few weeks. 

If you want to try out the fix for this, you can install

```bash
$ npm install @sinclair/typebox@0.32.0-dev-23
```
If you find any issues with the fix (noting some of the challenges getting these transforms correct) can investigate prior to publishing 0.32.0.
Cheers
S

**Comment by EmilEriksen** [comment]
@sinclairzx81 Hi. No worries at all. Thank you for solving it so quickly. 0.32.0 looks great. There are several other features I'm  going to use. Thank you for your fantastic work on this library!

**Comment by sinclairzx81** [comment]
@EmilEriksen Hi!

This fix should be ready on 0.32.0. Have just published the revision which you can install with the following.

```typescript
npm install @sinclair@typebox@0.32.0
```
Will close off this issue for now, but if you come across any other issues with Transforms, please let me know (they are a continuing endeavor to get right!)

All the best
S
